File: capacitor-plugin/android/src/main/java/it/hubagency/spatialscan/RoomExportData.kt
package it.hubagency.spatialscan

import kotlin.math.abs

/**
 * Export-ready representation of a room derived from RoomModel.
 * Bridges RoomModel (with WallModel + openings) to the export layer.
 * All wall Y coordinates are 0. Openings preserve full geometry.
 */

data class ExportOpening(
    val kind:                 OpeningKind,
    val offsetAlongWall:      Float,
    val width:                Float,
    val bottom:               Float,
    val height:               Float,
    // ── Connection metadata (propagated from OpeningModel) ────────────────────
    val isInternalConnection: Boolean = false,
    val linkedRoomId:         String? = null,
    val linkedOpeningId:      String? = null,
    val connectionLabel:      String? = null
)

data class ExportWall(
    val id:       String,
    val startX:   Float,
    val startZ:   Float,
    val endX:     Float,
    val endZ:     Float,
    val length:   Float,
    val height:   Float,
    val normalX:  Float,
    val normalZ:  Float,
    val dirX:     Float,
    val dirZ:     Float,
    val openings: List<ExportOpening>
)

class RoomExportData(
    val walls:      List<ExportWall>,
    val dimensions: RoomDimensions
) {
    companion object {
        fun fromRoomModel(rm: RoomModel): RoomExportData {
            val exportWalls = rm.walls.map { wm ->
                ExportWall(
                    id       = wm.id,
                    startX   = wm.start[0],
                    startZ   = wm.start[2],
                    endX     = wm.end[0],
                    endZ     = wm.end[2],
                    length   = wm.length,
                    height   = wm.height,
                    normalX  = wm.dirZ,       // outward normal: rotate dir 90° right
                    normalZ  = -wm.dirX,
                    dirX     = wm.dirX,
                    dirZ     = wm.dirZ,
                    openings = wm.openings.map { o ->
                        ExportOpening(
                            kind                 = o.kind,
                            offsetAlongWall      = o.offsetAlongWall,
                            width                = o.width,
                            bottom               = o.bottom,
                            height               = o.height,
                            isInternalConnection = o.isInternalConnection,
                            linkedRoomId         = o.linkedRoomId,
                            linkedOpeningId      = o.linkedOpeningId,
                            connectionLabel      = o.connectionLabel
                        )
                    }
                )
            }
            val dims = computeDimensions(exportWalls, rm.wallHeight)
            return RoomExportData(exportWalls, dims)
        }

        private fun computeDimensions(walls: List<ExportWall>, wallHeight: Float): RoomDimensions {
            if (walls.isEmpty()) return RoomDimensions(0.0, 0.0, wallHeight.toDouble(), 0.0, 0.0)
            val allX = walls.flatMap { listOf(it.startX.toDouble(), it.endX.toDouble()) }
            val allZ = walls.flatMap { listOf(it.startZ.toDouble(), it.endZ.toDouble()) }
            val width     = (allX.max() - allX.min()).coerceAtLeast(0.0)
            val length    = (allZ.max() - allZ.min()).coerceAtLeast(0.0)
            val perimeter = walls.sumOf { it.length.toDouble() } * 0.5  // preserve existing behaviour
            val area      = shoelace(walls)
            return RoomDimensions(width, length, wallHeight.toDouble(), area, perimeter)
        }

        /** Signed area via shoelace, walls form a closed polygon. */
        private fun shoelace(walls: List<ExportWall>): Double {
            var sum = 0.0
            for (w in walls) {
                sum += w.startX.toDouble() * w.endZ.toDouble() -
                       w.endX.toDouble()   * w.startZ.toDouble()
            }
            return abs(sum) / 2.0
        }
    }
}

File: ios/Plugin/Models/RoomExportData.swift
import Foundation

struct RoomDimensions {
    let width:     Double
    let length:    Double
    let height:    Double
    let area:      Double
    let perimeter: Double
}

struct ExportOpening {
    let kind:            OpeningKind
    let offsetAlongWall: Float
    let width:           Float
    let bottom:          Float
    let height:          Float
}

struct ExportWall {
    let id:       String
    let startX:   Float
    let startZ:   Float
    let endX:     Float
    let endZ:     Float
    let length:   Float
    let height:   Float
    let normalX:  Float
    let normalZ:  Float
    let dirX:     Float
    let dirZ:     Float
    let openings: [ExportOpening]
}

struct RoomExportData {
    let walls:        [ExportWall]
    let dimensions:   RoomDimensions
    let roomPolygons: [(String, [(Float, Float)])]

    init(walls: [ExportWall], dimensions: RoomDimensions,
         roomPolygons: [(String, [(Float, Float)])] = []) {
        self.walls = walls
        self.dimensions = dimensions
        self.roomPolygons = roomPolygons
    }

    static func fromRoomModel(_ rm: RoomModel) -> RoomExportData {
        let exportWalls = rm.walls.map { wm -> ExportWall in
            ExportWall(
                id: wm.id,
                startX: wm.start.x, startZ: wm.start.z,
                endX:   wm.end.x,   endZ:   wm.end.z,
                length: wm.length, height: wm.height,
                normalX: wm.dirZ, normalZ: -wm.dirX,
                dirX: wm.dirX, dirZ: wm.dirZ,
                openings: wm.openings.map { o in
                    ExportOpening(kind: o.kind, offsetAlongWall: o.offsetAlongWall,
                                  width: o.width, bottom: o.bottom, height: o.height)
                }
            )
        }
        let dims = computeDimensions(exportWalls, wallHeight: rm.wallHeight)
        return RoomExportData(walls: exportWalls, dimensions: dims)
    }

    private static func computeDimensions(_ walls: [ExportWall], wallHeight: Float) -> RoomDimensions {
        guard !walls.isEmpty else {
            return RoomDimensions(width: 0, length: 0, height: Double(wallHeight), area: 0, perimeter: 0)
        }
        let allX = walls.flatMap { [Double($0.startX), Double($0.endX)] }
        let allZ = walls.flatMap { [Double($0.startZ), Double($0.endZ)] }
        let width     = max(allX.max()! - allX.min()!, 0)
        let length    = max(allZ.max()! - allZ.min()!, 0)
        let perimeter = walls.reduce(0.0) { $0 + Double($1.length) } * 0.5
        let area      = shoelace(walls)
        return RoomDimensions(width: width, length: length,
                              height: Double(wallHeight), area: area, perimeter: perimeter)
    }

    private static func shoelace(_ walls: [ExportWall]) -> Double {
        walls.reduce(0.0) { sum, w in
            sum + Double(w.startX) * Double(w.endZ) - Double(w.endX) * Double(w.startZ)
        }.magnitude / 2.0
    }
}
